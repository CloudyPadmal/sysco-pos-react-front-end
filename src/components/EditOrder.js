import React, { Component } from 'react';
import { withRouter, Redirect } from 'react-router-dom'
import ListItemInEditOrder from './ListItemInEditOrder';

class EditOrder extends Component {

    constructor(props) {
        super(props);
        this.state = {
            UPDATEDITEMS: {},
            ITEMQUANTITYINORDER: {}
        };
    }

    // Once the component is mounted, filter out the productID and quantity of
    // the items in order and append it to local state
    componentDidMount() {
        let { items } = this.props.CURRENTORDER;
        let ITEMQUANTITYINORDER = {};
        for (var i in items) {
            ITEMQUANTITYINORDER[items[i].productID] = items[i].quantity;
        }
        this.setState({ ITEMQUANTITYINORDER });
    }

    /**************************************************************************
     * 
     *************************************************************************/
    UPDATE_THIS_ORDER = () => {
        let { UPDATEDITEMS } = this.state;
        let REFINEDITEMS = {};
        for (var ID in UPDATEDITEMS) {
            if (parseInt(UPDATEDITEMS[ID]) !== 0) {
                REFINEDITEMS[ID] = UPDATEDITEMS[ID];
            }
        }
        // TODO: Calculate differences
        this.props.UPDATE_ITEMS_IN_THIS_ORDER(REFINEDITEMS);
        setTimeout(() => {
            this.props.history.push('/my_orders')
        }, (50 * REFINEDITEMS.length));
    }

    /**************************************************************************
     * Do not make any modifications to the current order or the order list,
     * just redirect user back to order list
     *************************************************************************/
    CANCEL_EDITING_THE_ORDER = () => {
        this.props.history.push('/my_orders');
    }

    /**************************************************************************
     * Local function to add a single item to a temporary list of items to be 
     * added in to the order being updated. This will take care of duplicate 
     * entries but not zero entries. We'll need to refine that when updating
     * @param ID productID of the item being added
     * @param VALUE quantity of the item added
     *************************************************************************/
    ADD_THIS_ITEM_TO_ORDER = (ID, VALUE) => {
        try {
            delete this.state.UPDATEDITEMS.id;
        } catch (e) { }
        let tempAddedItems = this.state.UPDATEDITEMS;
        tempAddedItems[ID] = VALUE;
        this.setState({
            addedItems: tempAddedItems
        });
    }

    render() {
        if (this.props.CURRENTORDERID === "") {
            return (
                <Redirect to="/my_orders" />
            );
        }
        return (
            <div className="card" style={{ margin: '25px' }}>
                <div className="card-body">
                    <h5 className="card-title" style={{ margin: '0.5em 1em' }}>Editing Order {this.props.CURRENTORDERID}</h5>
                    {this.props.ITEMSLIST.map((item) => (
                        <ListItemInEditOrder
                            key={item._id}
                            ITEM={item}
                            keyID={item.productID}
                            ADD_THIS_ITEM_TO_ORDER={this.ADD_THIS_ITEM_TO_ORDER}
                            ITEMQTYINORDER={this.state.ITEMQUANTITYINORDER[item.productID]} />
                    ))}
                </div>
                <div className="card-body button-section" style={{ paddingTop: '0px' }}>
                    <div className="row">
                        <div className="col-12 d-flex justify-content-end">
                            <button onClick={this.UPDATE_THIS_ORDER} className="btn btn-primary" style={{ marginRight: '10px' }}>Update Order</button>
                            <button onClick={this.CANCEL_EDITING_THE_ORDER} className="btn btn-danger">Cancel</button>
                        </div>
                    </div>
                </div>
            </div>
        );
    }
}

export default withRouter(EditOrder);
