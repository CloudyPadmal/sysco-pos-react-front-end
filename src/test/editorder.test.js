import React from 'react';
import { shallowWrap, mountWrap } from './enzyme';

import EditOrder from '../components/EditOrder';

describe('Create Order', () => {

    it('displays a list of items to edit order', () => {
        const wrapper = mountWrap(<EditOrder
            ISLOGGEDIN={true}
            CURRENTORDERID={124134234}
            CURRENTORDER={{ _id: 23346, items: [] }}
            ITEMQUANTITY={{}}
            CLONEITEMQUANTITY={{}}
            TOTAL={123123.22}
            ITEMSLIST={[]}
            SET_THIS_ORDER_AS_CURRENT={() => { }}
            UPDATE_ITEMS_IN_THIS_ORDER={() => { }}
            DELETE_THIS_ITEM={() => { }}
            CLEAR_ORDER_UPDATE_PROCESS={() => { }}
            ADD_THIS_ITEM_TO_ITEMQUANTITY={() => { }}
            INDECCREMENT_ITEM_COUNT={() => { }} />);
        expect(wrapper).toMatchSnapshot();
    });

    it('redirects to login page if not logged in', () => {
        const wrapper = shallowWrap(<EditOrder
            ISLOGGEDIN={false}
            CURRENTORDERID={124134234}
            CURRENTORDER={{ _id: 23346, items: [] }}
            ITEMQUANTITY={{}}
            CLONEITEMQUANTITY={{}}
            TOTAL={123123.22}
            ITEMSLIST={{}}
            SET_THIS_ORDER_AS_CURRENT={() => { }}
            UPDATE_ITEMS_IN_THIS_ORDER={() => { }}
            DELETE_THIS_ITEM={() => { }}
            CLEAR_ORDER_UPDATE_PROCESS={() => { }}
            ADD_THIS_ITEM_TO_ITEMQUANTITY={() => { }}
            INDECCREMENT_ITEM_COUNT={() => { }} />);
        expect(wrapper).toMatchSnapshot();
    });
});