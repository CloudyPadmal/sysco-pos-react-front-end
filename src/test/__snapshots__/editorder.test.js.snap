// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Create Order displays a list of items to edit order 1`] = `
<withRouter(EditOrder)
  ADD_THIS_ITEM_TO_ITEMQUANTITY={[Function]}
  CLEAR_ORDER_UPDATE_PROCESS={[Function]}
  CLONEITEMQUANTITY={Object {}}
  CURRENTORDER={
    Object {
      "_id": 23346,
      "items": Array [],
    }
  }
  CURRENTORDERID={124134234}
  DELETE_THIS_ITEM={[Function]}
  INDECCREMENT_ITEM_COUNT={[Function]}
  ISLOGGEDIN={true}
  ITEMQUANTITY={Object {}}
  ITEMSLIST={Array []}
  SET_THIS_ORDER_AS_CURRENT={[Function]}
  TOTAL={123123.22}
  UPDATE_ITEMS_IN_THIS_ORDER={[Function]}
>
  <Route>
    <EditOrder
      ADD_THIS_ITEM_TO_ITEMQUANTITY={[Function]}
      CLEAR_ORDER_UPDATE_PROCESS={[Function]}
      CLONEITEMQUANTITY={Object {}}
      CURRENTORDER={
        Object {
          "_id": 23346,
          "items": Array [],
        }
      }
      CURRENTORDERID={124134234}
      DELETE_THIS_ITEM={[Function]}
      INDECCREMENT_ITEM_COUNT={[Function]}
      ISLOGGEDIN={true}
      ITEMQUANTITY={Object {}}
      ITEMSLIST={Array []}
      SET_THIS_ORDER_AS_CURRENT={[Function]}
      TOTAL={123123.22}
      UPDATE_ITEMS_IN_THIS_ORDER={[Function]}
      history={
        Object {
          "action": "POP",
          "block": [Function],
          "createHref": [Function],
          "go": [Function],
          "goBack": [Function],
          "goForward": [Function],
          "length": 1,
          "listen": [Function],
          "location": Object {
            "hash": "",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
          "push": [Function],
          "replace": [Function],
        }
      }
      location={Object {}}
      match={Object {}}
    >
      <div>
        <Beforeunload
          onBeforeunload={[Function]}
        />
        <div
          className="card"
          style={
            Object {
              "margin": "25px",
              "paddingBottom": "50px",
            }
          }
        >
          <div
            className="card-body"
          >
            <h5
              className="card-title"
              style={
                Object {
                  "margin": "0.5em 1em",
                }
              }
            >
              Editing Order 
              124134234
            </h5>
          </div>
        </div>
        <nav
          className="navbar navbar-light bg-light"
          style={
            Object {
              "bottom": "0",
              "overflow": "hidden",
              "paddingRight": "25px",
              "position": "fixed",
              "width": "100%",
              "zIndex": 99,
            }
          }
        >
          <div
            className="card-body button-section"
            style={
              Object {
                "paddingTop": "0px",
              }
            }
          >
            <div
              className="row"
            >
              <div
                className="col-4 d-flex justify-content-start d-inline"
              >
                <div
                  className="input-group-prepend"
                  style={
                    Object {
                      "marginLeft": "10px",
                    }
                  }
                >
                  <span
                    className="input-group-text"
                    style={
                      Object {
                        "borderBottomRightRadius": "0px",
                        "borderTopRightRadius": "0px",
                      }
                    }
                  >
                    <strong>
                      Total
                    </strong>
                  </span>
                  <span
                    className="input-group-text"
                    style={
                      Object {
                        "borderBottomLeftRadius": "0px",
                        "borderTopLeftRadius": "0px",
                      }
                    }
                  >
                    Rs. 
                    123123.22
                  </span>
                </div>
              </div>
              <div
                className="col-8 d-flex justify-content-end"
              >
                <button
                  className="btn btn-primary"
                  onClick={[Function]}
                  style={
                    Object {
                      "marginRight": "10px",
                    }
                  }
                >
                  <i
                    className="fas fa-pen"
                  />
                   Update Order
                </button>
                <button
                  className="btn btn-danger"
                  onClick={[Function]}
                >
                  <i
                    className="fas fa-times-circle"
                  />
                   Cancel
                </button>
              </div>
            </div>
          </div>
        </nav>
      </div>
    </EditOrder>
  </Route>
</withRouter(EditOrder)>
`;

exports[`Create Order redirects to login page if not logged in 1`] = `
<Route>
  <Component />
</Route>
`;
